lib/vdc/UnstructuredGrid3D.cpp

```diff
@@ -0,0 +1,236 @@
+#include <iostream>
+#include <vector>
+#include "vapor/VAssert.h"
+#include <cmath>
+#include <time.h>
+#ifdef Darwin
+    #include <mach/mach_time.h>
+#endif
+#ifdef _WINDOWS
+    #include "windows.h"
+    #include "Winbase.h"
+    #include <limits>
+#endif
+
+#include <vapor/utils.h>
+#include <vapor/vizutil.h>
+#include <vapor/UnstructuredGrid3D.h>
+
+using namespace std;
+using namespace VAPoR;
+
+UnstructuredGrid3D::UnstructuredGrid3D(const std::vector<size_t> &vertexDims, const std::vector<size_t> &faceDims, const std::vector<size_t> &edgeDims, const std::vector<size_t> &bs,
+                                       const std::vector<float *> &blks, const int *vertexOnFace, const int *faceOnVertex, const int *faceOnFace,
+                                       Location location,    // node,face, edge
+                                       size_t maxVertexPerFace, size_t maxFacePerVertex, long nodeOffset, long cellOffset, const UnstructuredGridCoordless &xug, const UnstructuredGridCoordless &yug,
+                                       const UnstructuredGridCoordless &zug)
+: UnstructuredGrid(vertexDims, faceDims, edgeDims, bs, blks, 3, vertexOnFace, faceOnVertex, faceOnFace, location, maxVertexPerFace, maxFacePerVertex, nodeOffset, cellOffset), _xug(xug), _yug(yug),
+  _zug(zug)
+{
+    VAssert(xug.GetDimensions().size() == 1);
+    VAssert(yug.GetDimensions().size() == 1);
+    VAssert(zug.GetDimensions().size() == 1);
+
+    VAssert(location == NODE);
+}
+
+
+vector<size_t> UnstructuredGrid3D::GetCoordDimensions(size_t dim) const
+{
+    if (dim == 0) {
+        return _xug.GetDimensions();
+    } else if (dim == 1) {
+        return _yug.GetDimensions();
+    } else if (dim == 2) {
+        return _zug.GetDimensions();
+    } else {
+        return (vector<size_t>(1, 1));
+    }
+}
+
+
+size_t UnstructuredGrid3D::GetGeometryDim() const { return (3); }
+
+
+void UnstructuredGrid3D::GetUserExtentsHelper(DblArr3 &minu, DblArr3 &maxu) const
+{
+    float range[2];
+
+    _xug.GetRange(range);
+    minu[0] = range[0];
+    maxu[0] = range[1];
+
+    _yug.GetRange(range);
+    minu[1] = range[0];
+    maxu[1] = range[1];
+
+    _zug.GetRange(range);
+    minu[2] = range[0];
+    maxu[2] = range[1];
+}
+
+
+void UnstructuredGrid3D::GetBoundingBox(const Size_tArr3 &min, const Size_tArr3 &max, DblArr3 &minu, DblArr3 &maxu) const
+{
+    Size_tArr3 cMin;
+    ClampIndex(min, cMin);
+
+    Size_tArr3 cMax;
+    ClampIndex(max, cMax);
+
+    float range[2];
+
+    _xug.GetRange(min, max, range);
+    minu[0] = range[0];
+    maxu[0] = range[1];
+
+    _yug.GetRange(min, max, range);
+    minu[1] = range[0];
+    maxu[1] = range[1];
+
+    _zug.GetRange(min, max, range);
+    minu[2] = range[0];
+    maxu[2] = range[1];
+
+    //    printf("UnstructuredGrid3D::%s(-, -)\n", __func__);
+    //    printf("\t min = [%f, %f, %f]\n", minu[0], minu[1], minu[2]);
+    //    printf("\t max = [%f, %f, %f]\n", maxu[0], maxu[1], maxu[2]);
```

Comment:
Remove debug code?

Reply:
I would like to leave this for later work.